rmf_site_editor/src/site/model.rs

```diff
@@ -16,528 +16,562 @@
 */
 
 use crate::{
-    interaction::{DragPlaneBundle, Preview, Selectable, MODEL_PREVIEW_LAYER},
-    site::{Category, Dependents, PreventDeletion, SiteAssets},
+    interaction::{DragPlaneBundle, Preview, MODEL_PREVIEW_LAYER},
+    site::SiteAssets,
     site_asset_io::MODEL_ENVIRONMENT_VARIABLE,
 };
 use bevy::{
-    asset::{LoadState, LoadedUntypedAsset},
-    gltf::Gltf,
-    prelude::*,
-    render::view::RenderLayers,
+    ecs::system::Command, gltf::Gltf, prelude::*, render::view::RenderLayers, scene::SceneInstance,
 };
+use bevy_impulse::*;
 use bevy_mod_outline::OutlineMeshExt;
-use rmf_site_format::{AssetSource, ModelMarker, Pending, Pose, Scale};
+use rmf_site_format::{AssetSource, ModelMarker, Pending, Scale};
 use smallvec::SmallVec;
-use std::any::TypeId;
+use std::{any::TypeId, future::Future};
+use thiserror::Error;
 
+/// Denotes the properties of the current spawned scene for the model, to despawn when updating AssetSource
+/// and avoid spurious reloading if the new `AssetSource` is equal to the old one.
+/// Models without a `ModelScene` either failed spawning or are still being spawned.
 #[derive(Component, Debug, Clone)]
 pub struct ModelScene {
     source: AssetSource,
-    format: TentativeModelFormat,
-    entity: Option<Entity>,
+    entity: Entity,
```

Comment:
Also is there any value in this struct being `pub`? If we want downstream users to have access to this then presumably we should provide some `pub` methods with `impl ModelScene` to give viewing access to its fields.

If we don't want its inner information to be exposed downstream then maybe we should make this private.

Reply:
The intention behind it (after this refactor) was stated in its docstring, it became a public API meant to only be used in queries through `With` or `Without` so downstream users can know whether the model was successfully spawned or not. This is now not needed anymore since there is a `ModelLoadingError` and we can use it instead.
However I tried to remove it and I got a bunch of compile warnings when the type is used in public APIs, two of them are just workflows and I believe they can be made private but one of them is indeed a [public system](https://github.com/open-rmf/rmf_site/blob/915edaddf24448e30ad5df60ca3e598b8fb00e0f/rmf_site_editor/src/site/model.rs#L471-L480) that can also be used (and indeed is) by downstream users.

So the only ways to make the type private to my knowledge would be either:

* Changing the system so it doesn't access it (maybe it could set scale to the entity with `ModelMarker` instead, but that's a slippery slope if we want to add any other children to it since transform propagation will scale them as well).
* Wrap the system in a plugin so we can keep it private and just export the plugin